src/src/Helpers/TarStream.cpp

```diff
@@ -0,0 +1,603 @@
+
+        if (_writePosition == _filesList[_fileIndex].tarSize) {
+          _streamState = TarStreamState_e::WritingHeader;
+          # if TAR_STREAM_DEBUG
+          addLog(LOG_LEVEL_INFO, concat(F("TarStream: Switch from "), F("WritingSlack to WritingHeader")));
+          # endif // if TAR_STREAM_DEBUG
+          _headerPosition = 0;
+        }
+        break;
+      }
+      case TarStreamState_e::WritingFinal:
+      {
+        // FIXME check entire 512 byte block or just ignore?
+        size_t toMove = std::min(size - bufOffset, TAR_HEADER_SIZE - _headerPosition);
+        memcpy(&_tarData[_headerPosition], &buf[bufOffset], toMove);
+        _headerPosition += toMove;
+        bufOffset       += toMove;
+        bool isBlank = true;
+
+        if (_headerPosition == TAR_HEADER_SIZE) {
+          for (size_t n = 0; n < TAR_HEADER_SIZE && isBlank; n++) {
+            isBlank &= (_tarData[n] == 0u);
+          }
+          _streamState = TarStreamState_e::WritingDone;
+          # if TAR_STREAM_DEBUG
+          addLog(LOG_LEVEL_INFO,
+                 strformat(F("%sWritingFinal to WritingDone, isBlank: %d"), String(F("TarStream: Switch from ")).c_str(), isBlank));
+          # endif // if TAR_STREAM_DEBUG
+        }
+        break;
+      }
+      case TarStreamState_e::WritingDone:
+      {
+        size_t toSkip = size - bufOffset;
+        bufOffset += toSkip;
+        # if TAR_STREAM_DEBUG
+        addLog(LOG_LEVEL_INFO, strformat(F("%sWritingDone, skipping: %d"), String(F("TarStream: ")).c_str(), toSkip));
+        # endif // if TAR_STREAM_DEBUG
+        break;
+      }
+      case TarStreamState_e::ReadingHeader: // Not here
+      case TarStreamState_e::ReadingFile:
+      case TarStreamState_e::ReadingSlack:
+      case TarStreamState_e::ReadingFinal:
+        break;
+      case TarStreamState_e::Error: // No real error state
+        break;
+    }
+
+    if (bufOffset < size) { // We got leftover bytes
+      stayInLoop = true;
+    }
+  }
+  delay(0);
+  _tarSize += size;
+  return size;
+}
+
+int TarStream::available() {
+  return _tarRemaining;
+}
+
+void TarStream::clearHeader() {
+  memset(&_tarHeader, 0, TAR_HEADER_SIZE);
+}
+
+void TarStream::setupHeader() {
+  clearHeader();
+  safe_strncpy(_tarHeader.name, _currentIndex.fileName.c_str(), sizeof(_tarHeader.name));
+  sprintf(_tarHeader.mode,    "%07o",  /*0777); //*/ TUREAD + TUWRITE + TUEXEC + TGREAD + TGWRITE + TGEXEC + TOREAD + TOWRITE + TOEXEC);
+  sprintf(_tarHeader.uid,     "%07o",  0);
+  sprintf(_tarHeader.gid,     "%07o",  0);
+  sprintf(_tarHeader.size,    "%011o", _currentIndex.fileSize);
+  sprintf(_tarHeader.mtime,   "%011o", node_time.getUnixTime()); // We don't have file-date/times, use current date/time
+  _tarHeader.typeflag = REGTYPE;
+  sprintf(_tarHeader.magic,   "%s",    TMAGIC);
+  sprintf(_tarHeader.version, "%s",    TVERSION);
+
+  for (size_t c = 0; c < sizeof(_tarHeader.chksum); c++) { _tarHeader.chksum[c] = ' '; }
```

Comment:
Watch out for `sizeof()` on arrays.
It is really tricky.
Best way is to do something like this, as in VScode you can then hover over the `constexpr` variable to see if you get a correct/expected value:

```c++
constexpr size_t chksum_size = sizeof(_tarHeader.chksum) / sizeof(_tarHeader.chksum[0]);
```
This way the compiler will turn this into a number during compile time.
And you will get the nr of elements as `sizeof()` of an array returns the number of bytes. So you need to divide it by the sizeof a single element.